BR should report error not panic if gc safepoint ts excess backup ts before running backup/restore task

1. What did you do?
If possible, provide a recipe for reproducing the error.

Backup with a ts that is less than gc safepoint. 

2. What did you expect to see?

BR reports error instead of running the backup task.

3. What did you see instead?

BR panics.

```
panic: cannot pass gc safe point check, aborting

goroutine 224 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000adbd40, 0xc000d0f780, 0x2, 0x2)
go.uber.org/zap@v1.16.0/zapcore/entry.go:234 +0x567
go.uber.org/zap.(*Logger).Panic(0xc0016fb6e0, 0x23a4105, 0x29, 0xc000d0f780, 0x2, 0x2)
go.uber.org/zap@v1.16.0/logger.go:226 +0x7f
github.com/pingcap/log.Panic(0x23a4105, 0x29, 0xc000d0f780, 0x2, 0x2)
github.com/pingcap/log@v0.0.0-20201112100606-8f1e84a3abc8/global.go:50 +0xff
github.com/pingcap/br/pkg/utils.StartServiceSafePointKeeper.func2(0x288e4a0, 0xc000226440)
github.com/pingcap/br@/pkg/utils/safe_point.go:111 +0x358
github.com/pingcap/br/pkg/utils.StartServiceSafePointKeeper.func3(0xc000c204b0, 0xc000c20500, 0x288e4a0, 0xc000226440, 0xc00002b380, 0xc00002b680)
github.com/pingcap/br@/pkg/utils/safe_point.go:131 +0x1b5
created by github.com/pingcap/br/pkg/utils.StartServiceSafePointKeeper
github.com/pingcap/br@/pkg/utils/safe_point.go:120 +0x209
```

--- a/pkg/task/backup.go
+++ b/pkg/task/backup.go
@@ -232,9 +232,11 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
 		sp.BackupTS = cfg.LastBackupTS
 	}
 
-	log.Info("current backup safePoint job",
-		zap.Object("safePoint", sp))
-	utils.StartServiceSafePointKeeper(ctx, mgr.GetPDClient(), sp)
+	log.Info("current backup safePoint job", zap.Object("safePoint", sp))
+	err = utils.StartServiceSafePointKeeper(ctx, mgr.GetPDClient(), sp)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	isIncrementalBackup := cfg.LastBackupTS > 0
 

--- a/pkg/task/restore.go
+++ b/pkg/task/restore.go
@@ -229,7 +229,10 @@ func RunRestore(c context.Context, g glue.Glue, cmdName string, cfg *RestoreConf
 	// restore checksum will check safe point with its start ts, see details at
 	// https://github.com/pingcap/tidb/blob/180c02127105bed73712050594da6ead4d70a85f/store/tikv/kv.go#L186-L190
 	// so, we should keep the safe point unchangeable. to avoid GC life time is shorter than transaction duration.
-	utils.StartServiceSafePointKeeper(ctx, mgr.GetPDClient(), sp)
+	err = utils.StartServiceSafePointKeeper(ctx, mgr.GetPDClient(), sp)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
 	var newTS uint64
 	if client.IsIncremental() {

--- a/pkg/utils/safe_point.go
+++ b/pkg/utils/safe_point.go
@@ -74,13 +74,11 @@ func CheckGCSafePoint(ctx context.Context, pdClient pd.Client, ts uint64) error 
 	return nil
 }
 
-// UpdateServiceSafePoint register BackupTS to PD, to lock down BackupTS as safePoint with TTL seconds.
-func UpdateServiceSafePoint(ctx context.Context, pdClient pd.Client, sp BRServiceSafePoint) error {
-	log.Debug("update PD safePoint limit with TTL",
-		zap.Object("safePoint", sp))
+// updateServiceSafePoint register BackupTS to PD, to lock down BackupTS as safePoint with TTL seconds.
+func updateServiceSafePoint(ctx context.Context, pdClient pd.Client, sp BRServiceSafePoint) error {
+	log.Debug("update PD safePoint limit with TTL", zap.Object("safePoint", sp))
 
-	lastSafePoint, err := pdClient.UpdateServiceGCSafePoint(ctx,
-		sp.ID, sp.TTL, sp.BackupTS-1)
+	lastSafePoint, err := pdClient.UpdateServiceGCSafePoint(ctx, sp.ID, sp.TTL, sp.BackupTS-1)
 	if lastSafePoint > sp.BackupTS-1 {
 		log.Warn("service GC safe point lost, we may fail to back up if GC lifetime isn't long enough",
 			zap.Uint64("lastSafePoint", lastSafePoint),
@@ -96,27 +94,23 @@ func StartServiceSafePointKeeper(
 	ctx context.Context,
 	pdClient pd.Client,
 	sp BRServiceSafePoint,
-) {
+) error {
+	if sp.ID == "" || sp.TTL <= 0 {
+		return errors.Annotatef(berrors.ErrInvalidArgument, "invalid service safe point %v", sp)
+	}
+	if err := CheckGCSafePoint(ctx, pdClient, sp.BackupTS); err != nil {
+		return errors.Trace(err)
+	}
+	// Update service safe point immediately to cover the gap between starting
+	// update goroutine and updating service safe point.
+	if err := updateServiceSafePoint(ctx, pdClient, sp); err != nil {
+		return errors.Trace(err)
+	}
+
 	// It would be OK since TTL won't be zero, so gapTime should > `0.
 	updateGapTime := time.Duration(sp.TTL) * time.Second / preUpdateServiceSafePointFactor
-	update := func(ctx context.Context) {
-		if err := UpdateServiceSafePoint(ctx, pdClient, sp); err != nil {
-			log.Warn("failed to update service safe point, backup may fail if gc triggered",
-				zap.Error(err),
-			)
-		}
-	}
-	check := func(ctx context.Context) {
-		if err := CheckGCSafePoint(ctx, pdClient, sp.BackupTS); err != nil {
-			log.Panic("cannot pass gc safe point check, aborting",
-				zap.Error(err),
-				zap.Object("safePoint", sp),
-			)
-		}
-	}
 	updateTick := time.NewTicker(updateGapTime)
 	checkTick := time.NewTicker(checkGCSafePointGapTime)
-	update(ctx)
 	go func() {
 		defer updateTick.Stop()
 		defer checkTick.Stop()
@@ -126,10 +120,20 @@ func StartServiceSafePointKeeper(
 				log.Debug("service safe point keeper exited")
 				return
 			case <-updateTick.C:
-				update(ctx)
+				if err := updateServiceSafePoint(ctx, pdClient, sp); err != nil {
+					log.Warn("failed to update service safe point, backup may fail if gc triggered",
+						zap.Error(err),
+					)
+				}
 			case <-checkTick.C:
-				check(ctx)
+				if err := CheckGCSafePoint(ctx, pdClient, sp.BackupTS); err != nil {
+					log.Panic("cannot pass gc safe point check, aborting",
+						zap.Error(err),
+						zap.Object("safePoint", sp),
+					)
+				}
 			}
 		}
 	}()
+	return nil
 }

--- a/pkg/utils/safe_point_test.go
+++ b/pkg/utils/safe_point_test.go
@@ -47,15 +47,93 @@ func (s *testSafePointSuite) TestCheckGCSafepoint(c *C) {
 type mockSafePoint struct {
 	sync.Mutex
 	pd.Client
-	safepoint uint64
+	safepoint           uint64
+	minServiceSafepoint uint64
+}
+
+func (m *mockSafePoint) UpdateServiceGCSafePoint(ctx context.Context, serviceID string, ttl int64, safePoint uint64) (uint64, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.safepoint > safePoint {
+		return m.safepoint, nil
+	}
+	if m.minServiceSafepoint == 0 || m.minServiceSafepoint > safePoint {
+		m.minServiceSafepoint = safePoint
+	}
+	return m.minServiceSafepoint, nil
 }
 
 func (m *mockSafePoint) UpdateGCSafePoint(ctx context.Context, safePoint uint64) (uint64, error) {
 	m.Lock()
 	defer m.Unlock()
 
-	if m.safepoint < safePoint {
+	if m.safepoint < safePoint && safePoint < m.minServiceSafepoint {
 		m.safepoint = safePoint
 	}
 	return m.safepoint, nil
+}
+
+func (s *testSafePointSuite) TestStartServiceSafePointKeeper(c *C) {
+	pdClient := &mockSafePoint{safepoint: 2333}
+
+	cases := []struct {
+		sp utils.BRServiceSafePoint
+		ok bool
+	}{
+		{
+			utils.BRServiceSafePoint{
+				ID:       "br",
+				TTL:      10,
+				BackupTS: 2333 + 1,
+			},
+			true,
+		},
+
+		// Invalid TTL.
+		{
+			utils.BRServiceSafePoint{
+				ID:       "br",
+				TTL:      0,
+				BackupTS: 2333 + 1,
+			}, false,
+		},
+
+		// Invalid ID.
+		{
+			utils.BRServiceSafePoint{
+				ID:       "",
+				TTL:      0,
+				BackupTS: 2333 + 1,
+			},
+			false,
+		},
+
+		// BackupTS is too small.
+		{
+			utils.BRServiceSafePoint{
+				ID:       "br",
+				TTL:      10,
+				BackupTS: 2333,
+			}, false,
+		},
+		{
+			utils.BRServiceSafePoint{
+				ID:       "br",
+				TTL:      10,
+				BackupTS: 2333 - 1,
+			},
+			false,
+		},
+	}
+	for i, cs := range cases {
+		ctx, cancel := context.WithCancel(context.Background())
+		err := utils.StartServiceSafePointKeeper(ctx, pdClient, cs.sp)
+		checker := IsNil
+		if !cs.ok {
+			checker = NotNil
+		}
+		c.Assert(err, checker, Commentf("case #%d, %v", i, cs))
+		cancel()
+	}
 }

